### DISCUSSION TITLE
Unable to generate / save PDF getting error 'node' is not recognized as an internal or external command, operable program or batch file

### DISCUSSION THREAD

**Discussion by salmanjaveed** [spatie/laravel-pdf#32] [Q&A]
A simple test file is not being saved as PDF. Keep getting the above error message.

`Pdf::view('support.test')->save('test.pdf');`

The view is a simple 1 sentence. Windows 11 / node installed and is on the path.

Any help on this would be great. The full error text follows:
`The command "node "C:\Users\xxx\Desktop\xx\vendor\spatie\browsershot\src/../bin/browser.cjs" "{""url"":""file:\/\/C:\\Users\\xxx\\AppData\\Local\\Temp\\1919060455-0013397001704897567\\index.html"",""action"":""pdf"",""options"":{""path"":""test.pdf"",""args"":[],""viewport"":{""width"":800,""height"":600},""displayHeaderFooter"":false,""printBackground"":true}}"" failed. Exit Code: 1(General error) Working directory: C:\Users\xxx\Desktop\xx\public Output: ================ Error Output: ================ 'node' is not recognized as an internal or external command, operable program or batch file.`

**Comment by TomOusselin** [comment]
It looks like i have a similar problem : 
![image](https://github.com/spatie/laravel-pdf/assets/76063513/9bf62163-4367-4152-baa3-c30d461b269c)

**Reply by BeniNasufi** [reply]
having same issue

**Comment by BeniNasufi** [comment]
Found a solution through browsershots forum. The problem is while running with php artisan serve
It will work if you open the project through Apache from xampp with localhost/project/path/to/function
Although was close to fix it from artisan but got stuck on an error which required to open chrome to get to that part with artisan you need to add 
```
  Pdf::view('pdf.str',compact('property','currency_sign'))
    ->withBrowsershot(function (Browsershot $browsershot) {
        $browsershot->setNodeBinary('C:/Program Files/nodejs/node')
            ->setNpmBinary('C:/Program Files/nodejs/npm')
            ->setChromePath("C:\Program Files\Google\Chrome\Application\chrome.exe")
            ->newHeadless();
    })
```
            
 this will fix the node not found exception and old headless implementation warning.
If you manage to get something please update it here otherwise it runs with Apache perfectly.

**Reply by firdavsibodullaev** [reply]
It didn't help to me, I have the same error, but all paths are right

![image](https://github.com/spatie/laravel-pdf/assets/59536101/ac1aeccd-4a2b-4189-a142-f53a9bfedd10)
I tried and forward slash and backslash, but it didn't help me

**Reply by sToiBe** [reply]
#This works for me

> `return Pdf::view('pdfs.contract')
>     ->withBrowsershot(function (Browsershot $browsershot) {
>         $browsershot->setNodeBinary('C:/Program Files/nodejs/node')
>             ->setNpmBinary('C:/Program Files/nodejs/npm')
>             ->setChromePath("C:\Program Files\Google\Chrome\Application\chrome.exe")
>             ->newHeadless();
>     })->inline();`

**Comment by FocusTvi** [comment]
Try using the disk() instead. 

Eg: 
```
Pdf::view($view, $data)
            ->withBrowsershot(function (Browsershot $browsershot) {
                $browsershot
                    ->setNodeBinary(config('app.pdf.node_binary_path'))
                    ->setNpmBinary(config('app.pdf.npm_binary_path'))
                    ->setIncludePath(config('app.pdf.include_path'))
                    ->setChromePath(config('app.pdf.chrome_path'))
                    ->setCustomTempPath(config('app.pdf.custom_temp_path'))
                    ->waitUntilNetworkIdle();
            })
            ->disk('custom_local')
            ->save('file.pdf');
```

The disk should be in the filesystems.php 

```
 'disks' => [
        'custom_local' => [
            'driver' => 'local',
            'root'   => storage_path('app/custom_local'),
        ],
```

**Comment by alamgir-napollo** [comment]
public function generatePDF()
    {
        $data = [
            'invoiceNumber' => 'INV-2025-07-001',
            'date' => date('Y-m-d'),
            'customerName' => 'John Doe',
            'customerAddress' => '123 Main St, Anytown, USA',
            'items' => [
                ['name' => 'Product A', 'quantity' => 2, 'price' => 10.50],
                ['name' => 'Service B', 'quantity' => 1, 'price' => 50.00],
                ['name' => 'Product C', 'quantity' => 3, 'price' => 5.25],
            ],
            'total' => 0,
        ];

        foreach ($data['items'] as $item) {
            $data['total'] += ($item['quantity'] * $item['price']);
        }

        $html = View::make('invoice', $data)->render();

        try {
            $pdfPath = public_path(time().'_file.pdf');

            Browsershot::html($html)
                ->showBackground()
                ->format('A4')
                ->margins(10, 10, 10, 10)
                ->save($pdfPath);

            return response()->json([
                'message' => 'PDF generated and saved successfully!',
                'path' => $pdfPath
            ]);

        } catch (\Exception $e) {
            return response()->json([
                'error' => 'Failed to generate PDF: ' . $e->getMessage()
            ], 500);
        }

    }
    
    The above code works fine with css 3 also
